app/src/services/cpk.ts

```diff
@@ -0,0 +1,124 @@
+import { ethers, Wallet } from 'ethers'
+import CPK from 'contract-proxy-kit'
+
+import { getLogger } from '../util/logger'
+import { ConditionalTokenService, ERC20Service, MarketMakerService } from './index'
+import { BigNumber } from 'ethers/utils'
+
+const logger = getLogger('Services::CPKService')
+
+interface CPKBuyOutcomesParams {
+  provider: any
+  cost: BigNumber
+  amount: BigNumber
+  outcomeIndex: number
+  marketMaker: MarketMakerService
+  conditionalTokens: ConditionalTokenService
+}
+
+class CPKService {
+  static buyOutcomes = async ({
+    provider,
+    cost,
+    amount,
+    outcomeIndex,
+    marketMaker,
+    conditionalTokens,
+  }: CPKBuyOutcomesParams) => {
+    try {
+      const signer: Wallet = provider.getSigner()
+      const account = await signer.getAddress()
+
+      const collateralAddress = await marketMaker.getCollateralToken()
+      const marketMakerAddress = marketMaker.address
+      const conditionalTokensAddress = conditionalTokens.address
+
+      const cpk = await CPK.create({ ethers, signer })
+      const cpkAddress = cpk.address
+
+      // Approve amount of collateral to the CPK
+      const collateralService = new ERC20Service(provider, account, collateralAddress)
+      const hasEnoughAlowance = await collateralService.hasEnoughAllowance(
+        account,
+        cpkAddress,
+        cost,
+      )
+
+      if (!hasEnoughAlowance) {
+        await collateralService.approveUnlimited(cpkAddress)
+      }
+
+      logger.log(`CPK address: ${cpkAddress}`)
+
+      const outcomeTokensToBuy = await marketMaker.calcBuyAmount(amount, outcomeIndex)
+      logger.log(`Min outcome tokens to buy: ${outcomeTokensToBuy}`)
+
+      const conditionId = await marketMaker.getConditionId()
+      const collectionId = await conditionalTokens.getCollectionIdForOutcome(
+        conditionId,
+        1 << outcomeIndex,
+      )
+      logger.debug(
+        `Collection ID for outcome index ${outcomeIndex} and condition id ${conditionId} : ${collectionId}`,
+      )
+      const positionIdForCollectionId = await conditionalTokens.getPositionId(
+        collateralAddress,
+        collectionId,
+      )
+
+      const txObject = await cpk.execTransactions(
+        [
+          // Step 1: Transfer an amount (cost) from the user to the CPK
+          {
+            operation: CPK.CALL,
+            to: collateralAddress,
+            value: 0,
+            data: ERC20Service.encodeTransferFrom(account, cpkAddress, cost),
+          },
+          // Step 2: Approve an amount(cost) to be used by the market maker (later that amount(cost) will be transferred to the market maker)
```

Comment:
This comment is wrong, the approval is unlimited.

Reply:
Very low priority but: the second time a purchase is done, the approval isn't necessary and therefore wasteful. This step should only be added if the allowance of the CPK to the market maker is not enough.

Reply:
Done!